driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java

```diff
@@ -306,18 +314,29 @@ void markAsPinned(final Connection connectionToPin, final Connection.PinningMode
 
         @Override
         void doClose() {
-            if (isSkipReleasingServerResourcesOnClose()) {
-                unsetServerCursor();
+            TimeoutContext timeoutContext = operationContext.getTimeoutContext();
+            timeoutContext.resetToDefaultMaxTime();
+            if (resetTimeoutWhenClosing) {
+                timeoutContext.doWithResetTimeout(this::releaseResources);
+            } else {
+                releaseResources();
             }
-            resetTimeout();
+        }
+
+        private void releaseResources() {
             try {
-                if (getServerCursor() != null) {
+                if (isSkipReleasingServerResourcesOnClose()) {
+                    unsetServerCursor();
+                }
+                if (super.getServerCursor() != null) {
                     Connection connection = getConnection();
                     try {
                         releaseServerResources(connection);
                     } finally {
                         connection.release();
                     }
+                } else {
+                    // do nothing
```

Comment:
Removed all such empty `else` blocks in 0ea7ccc0c7515c9586680789c61f86e4b5c0dc52.

We can't do `else if` because the two conditions that are currently checked with two `if`s are not mutually exclusive.